bundles/org.openhab.binding.revogi/src/main/java/org/openhab/binding/revogi/internal/udp/UdpSenderService.java

```diff
@@ -0,0 +1,135 @@
+        this.datagramSocketWrapper = datagramSocketWrapper;
+    }
+
+    public CompletableFuture<List<UdpResponse>> broadcastUpdDatagram(String content) {
+        List<String> allBroadcastAddresses = NetUtil.getAllBroadcastAddresses();
+        CompletableFuture<List<UdpResponse>> future = new CompletableFuture<>();
+        Executors.newCachedThreadPool().submit(() -> future.complete(allBroadcastAddresses.stream().map(address -> {
+            try {
+                return sendMessage(content, InetAddress.getByName(address));
+            } catch (UnknownHostException e) {
+                logger.warn("Could not find host with IP {}", address);
+                return new ArrayList<UdpResponse>();
+            }
+        }).flatMap(Collection::stream).distinct().collect(toList())));
+        return future;
+    }
+
+    public CompletableFuture<List<UdpResponse>> sendMessage(String content, String ipAddress) {
+        try {
+            CompletableFuture<List<UdpResponse>> future = new CompletableFuture<>();
+            InetAddress inetAddress = InetAddress.getByName(ipAddress);
+            Executors.newCachedThreadPool().submit(() -> future.complete(sendMessage(content, inetAddress)));
+            return future;
+        } catch (UnknownHostException e) {
+            logger.warn("Could not find host with IP {}", ipAddress);
+            return CompletableFuture.completedFuture(Collections.emptyList());
+        }
+    }
+
+    private List<UdpResponse> sendMessage(String content, InetAddress inetAddress) {
+        logger.debug("Using address {}", inetAddress);
+        byte[] buf = content.getBytes(Charset.defaultCharset());
+        DatagramPacket packet = new DatagramPacket(buf, buf.length, inetAddress, REVOGI_PORT);
+        List<UdpResponse> responses = Collections.emptyList();
+        try {
+            datagramSocketWrapper.initSocket();
+            datagramSocketWrapper.sendPacket(packet);
+            responses = getUdpResponses();
+        } catch (IOException e) {
+            logger.warn("Error sending message or reading anwser {}", e.getMessage(), e);
+        } finally {
+            datagramSocketWrapper.closeSocket();
+        }
+        return responses;
+    }
+
+    private List<UdpResponse> getUdpResponses() {
+        int timeoutCounter = 0;
+        List<UdpResponse> list = new ArrayList<>();
+        while (timeoutCounter < MAX_TIMEOUT_COUNT) {
+            byte[] receivedBuf = new byte[512];
+            DatagramPacket answer = new DatagramPacket(receivedBuf, receivedBuf.length);
+            try {
+                datagramSocketWrapper.receiveAnswer(answer);
+            } catch (SocketTimeoutException | SocketException e) {
+                timeoutCounter++;
+                logger.info("Socket receive time no. {}", timeoutCounter);
+                try {
+                    TimeUnit.MILLISECONDS.sleep(timeoutCounter * TIMEOUT_BASE_VALUE);
```

Comment:
When you change the threading to use the existing schedulers, better replace the sleep by scheduling another task to prevent blocking of the current task.

Reply:
I'm not sure what you mean what I should do. Somewhere I need to tell the software to wait a certain amount of time until it looks for udp responses again. I already refactored this part to not block the main thread.

Reply:
Your current implementation is blocking the current thread. You can avoid the blocking by using `scheduler.scheduleWithFixedDelay()`, which schedules the task after the specified time without blocking.

Reply:
I just saw, that the timeout is quite short. I don't see any problems if you keep it as is.